open_lm/train.py

```diff
@@ -132,14 +134,32 @@ def train_one_epoch(model, data, loss, epoch, step, optimizer, scaler, scheduler
 
     end = time.time()
 
+    try:
+        starting_step = int(optimizer.state_dict()["state"][0]["step"].item())
+    except KeyError:
+        # Throws keyerror if it is the first step
+        starting_step = 0
+    done_training = False
+
+    other_proc_ran_out_of_data = False
     for i, batch in enumerate(dataloader):
         if not args.skip_scheduler:
             scheduler(step)
 
-        if step > total_steps:
+        if step >= total_steps:
             logging.warning(f"step: {step} exceeds total_steps: {total_steps}. ending training.")
             break
 
+        # If some process no longer has data, stop training - for this epoch
+        procs_with_data = torch.tensor(1, dtype=torch.long, device=device)
```

Comment:
Right now we call dist.all_reduce here and then again at the end of the loop. This is a bit confusing and error-prone. Can we instead do the following?

1. Replace 
    ```
    for i, batch in enumerate(dataloader):
    ```
    with
    ```
    for i in itertools.count():
    ```
    to get an infinite iterator.
2. Add the following here:
    ```
    try:
        batch = next(dataloader)
        has_data = torch.tensor(1, dtype=torch.long, device=device)
    except StopIteration:
        has_data = torch.tensor(0, dtype=torch.long, device=device)
    dist.all_reduce(has_data, op=ReduceOp.SUM)
    if has_data < args.world_size:
        break  # Double check - not sure if we need to do anything else before breaking.
    ```
3. Remove dist.all_reduce call from after the loop.

Reply:
I agree, this is far cleaner to what I had before.

However, two calls to dist.all_reduce need to be made (since I'm also communicating the loss value globally, which needs to happen after making sure there is data for it).